File: helix_desktop/Helix_Desktop_Netduino/Helix_Desktop_Netduino/EvoController.cs
using System;
using Microsoft.SPOT;

namespace Helix
{
    class EvoController
    {
        public static GA GA = new GA();

        bool Play;
        int Step;

        //bool Selection;
        //bool Heredity;
        bool Mutation;
        bool MutationEvent;

        double current_mutation_rate;

        public EvoController()
        {
            Play = true;
            Step = 1;

            Mutation = true;
            MutationEvent = false;

            current_mutation_rate = GA.per_site_mutation_rate;
        }

        public void next()
        {
            if (Play)
            {
                if (MutationEvent)
                {
                    current_mutation_rate = GA.per_site_mutation_rate;
                    GA.per_site_mutation_rate = 0.5;
                }

                GA.evolve();

                if (MutationEvent)
                {
                    GA.per_site_mutation_rate = current_mutation_rate;
                    MutationEvent = false;
                }
            }
            else
            {
                if (Step > 0)
                {
                    GA.evolve();
                    Step--;
                }
            }
        }

        public void toggle_pauseplay()
        {
            Play = !Play;
        }

        public void mutation_event()
        {
            MutationEvent = true;
        }

        public void toggle_selection()
        {           
            GA.use_selection = !GA.use_selection;
        }

        public void toggle_mutation()
        {
            if (Mutation)
            {
                current_mutation_rate = GA.per_site_mutation_rate;
                GA.per_site_mutation_rate = 0;
                Mutation = false;
            }
            else
            {
                GA.per_site_mutation_rate = current_mutation_rate;
                Mutation = true;
            }
        }

        public void toggle_heredity()
        {
            GA.use_heredity = !GA.use_heredity;
        }


        public void take_step()
        {
            if (!Play) // this only makes sense in the context of being paused
                Step++;
        }

        // ceiling of 0.5
        public void raise_mut_rate()
        {
            if (Mutation && GA.per_site_mutation_rate < 0.5)
                GA.per_site_mutation_rate *= 2;
   
        }

        // floor of 0.01 
        public void lower_mut_rate()
        {
            if (Mutation && GA.per_site_mutation_rate > 0.01)
                GA.per_site_mutation_rate /= 2;
        }

        public void reset_ga() 
        {
            Play = true;
            Step = 1;
            GA.per_site_mutation_rate = 0.01;
            GA.use_heredity = true;
            GA.use_selection = true;
            GA.reset_population();
        }

        public int generation()
        {
            return GA.generation;
        }

        public double per_site_mutation_rate()
        {
            return GA.per_site_mutation_rate;
        }



    }
}

File: helix_fullscale/Helix/Helix/EvoController.cpp
//
//  EvoController.cpp
//  Helix
//
//  Created by Rosangela Canino-Koning on 1/28/13.
//  Copyright (c) 2013 Sube Studio. All rights reserved.
//

#include "EvoController.h"

EvoController::EvoController()
{
    Play = true;
    Step = 1;
    
    Mutation = true;
    MutationEvent = false;
    
    //current_mutation_rate = Engine.per_site_mutation_rate;
    
    valid_mutation_rates = {0.01, 0.02, 0.04, 0.08, 0.16, 0.32};
    valid_heritability_values = {0.0, 0.2, 0.4, 0.6, 0.8, 1.0};
//    valid_phenotypic_plasticities = {0, 5, 10, 20, 40, 80};
    valid_selection_strengths = {0.0, 0.2, 0.4, 0.6, 0.8, 1.0};
    
    mut_rate_setting = 0; // min
//    pheno_plast_setting = 0;
    heritability_setting = 5; // max
    selection_setting = 5; // max
    
    
}

void EvoController::init()
{
    Engine.init();
}

void EvoController::next()
{
    if (Play)
    {
        Engine.evolve();

    }
    else
    {
        if (Step > 0)
        {
            Engine.evolve();
            Step--;
        }
    }
}

void EvoController::toggle_pauseplay()
{
    Play = !Play;
}


void EvoController::take_step()
{
    if (!Play) // this only makes sense in the context of being paused
        Step++;
}

void EvoController::set_mutation_rate(int aVal)
{
    Engine.per_site_mutation_rate = valid_mutation_rates[aVal];
    mut_rate_setting = aVal;
}

int EvoController::get_mutation_rate_setting()
{
    return mut_rate_setting;
}

void EvoController::set_heritability(int aVal)
{
    Engine.heritability = valid_heritability_values[aVal];
    heritability_setting = aVal;
}

int EvoController::get_heritability_setting()
{
    return heritability_setting;
}

void EvoController::set_selection_strength(int aVal)
{
    Engine.selection_strength = valid_selection_strengths[aVal];
    selection_setting = aVal;
}

int EvoController::get_selection_strength_setting()
{
    return selection_setting;
}

void EvoController::reset_ga()
{
    Play = true;
    Step = 1;
    
    Engine.per_site_mutation_rate = __DEFAULT_MUTATION_RATE__;
    Engine.heritability = __DEFAULT_HERITABILITY__;
    Engine.selection_strength = __DEFAULT_SELECTION_STRENGTH__;
    
    mut_rate_setting = 0;
    heritability_setting = 5;
    selection_setting = 5;
    
    Engine.reset_ga();
}

int EvoController::generation()
{
    return Engine.generation;
}

double EvoController::per_site_mutation_rate()
{
    return Engine.per_site_mutation_rate;
}



